src/GraphQL/API/Enum.hs

```diff
@@ -0,0 +1,101 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DefaultSignatures #-}
+
+module GraphQL.API.Enum
+  ( GraphQLEnum(..)
+  ) where
+
+import Protolude hiding (Enum, U1, TypeError)
+import GraphQL.Internal.AST (Name, nameFromSymbol, NameError)
+import qualified GraphQL.Value as GValue
+import GHC.Generics (M1(..), Rep, Meta(..), C1, U1(..), D, (:+:)(..))
+import GHC.TypeLits (KnownSymbol, TypeError, ErrorMessage(..))
+
+
+class GenricEnumValues (f :: Type -> Type) where
+  genericEnumValues :: [Either NameError Name]
+  genericEnumFromValue :: GValue.Value -> Either Text (f p)
```

Comment:
Why does this take `Value`? Since the scope of this class is restricted to enums, it should take whatever type `x` is in `ValueEnum x`. 

Then our code that calls `enumToValue` and `enumFromValue` does the pattern match. This renders the "should have been caught at top-level" `panic` redundant. 

… oh, I see that's `Name`. Still `Name` would be better here than `Value` which is overly broad and makes it too easy for implementers to provide buggy implementations. Later (or even in this PR, actually), we can make an `newtype Enum = Enum Name` alias, which would also be useful later for providing that the enum isn't `true`, `false`, or `null`.